soroban-env-host/src/native_contract/token/metadata.rs

```diff
@@ -20,33 +24,149 @@ pub struct TokenMetadata {
 
 pub const DECIMAL: u32 = 7;
 
+// This does a specific and fairly unique escaping transformation as defined
+// in TxRep / SEP-0011.
+fn render_sep0011_asset_code(
+    buf: &[u8],
+    out: &mut std::string::String,
+) -> Result<(), ConversionError> {
+    if buf.len() != 4 && buf.len() != 12 {
+        return Err(ConversionError);
+    }
+    for (i, x) in buf.iter().enumerate() {
+        match *x {
+            // When dealing with a 4-byte asset code we stop at the first NUL.
+            0 if buf.len() == 4 => break,
+            // When dealing with a 12-byte asset code we continue escaping NULs
+            // as \x00 until past the 5th byte, so that the result is
+            // unambiguously different than a 4-byte code.
+            0 if buf.len() == 12 && i > 4 => break,
+            b':' | b'\\' | 0..=0x20 | 0x7f..=0xff => {
+                write!(out, r"\x{:02x}", x).map_err(|_| ConversionError)?
+            }
```

Comment:
Am I correct in thinking that this code will never get triggered because stellar-core never allows for the use of an asset that is a 12-byte with less than 5 non-null characters?

If I am incorrect in thinking that, then we probably have a second bug in that the token wrap method is allowing instantiation of contracts that wrap invalid assets.

Reply:
Line 43? It stops the loop at the first null after char 4 (the 5th char). I.e. if you have a 12-char code with 5 leading non-null chars like "ABCDE\0\0\0\0\0\0\0", the 5 chars from buf[0] through buf[4] will get converted, and then buf[5] which is null will be seen by line 43 of the loop and terminate the loop with just "ABCDE" in the result.

I'm not sure if that answers your question. It's not dead code as far as I know!

Reply:
(The reason it's conditionalized on 5 is so that a _shorter_ code like "ABC\0\0\0\0\0\0\0\0\0" will get converted to "ABC\x00\x00" rather than "ABC". This is how txrep appears to be differentiating literal 12-char codes from literal 4-char codes)

Reply:
My question is does the contract allow creating a token with "ABC\0\0" because my understanding is it shouldn't be possible to create a AssetCode12 asset with null bytes on Stellar.

Reply:
i.e. Does Soroban let us create contract instances for assets that can't exist on Stellar Classic.

Reply:
Oh, I have no idea. I'm just saying even if "ABC\0\0" isn't a legal asset, that line of code will get triggered when rendering "ABCDE" which is legal. The code's not dead. Or were you asking about rendering escaped nulls specifically, that they can never happen on a 12-char code? That might be true. The escape-rendering line is executed for lots of reasons though: trailing nulls in 4-char codes as well as any non-nulls outside the printable range allowed.

Going to your question about allowed assets: it looks vaguely to me like [a 12-char asset "ABC\0\0" is allowed in soroban](https://github.com/stellar/rs-soroban-env/blob/main/soroban-env-host/src/native_contract/token/contract.rs#L113-L182) but I don't really understand the setup path -- who can call that and how it gets validated before or after.  I guess you're looking for a check somewhere that mirrors [this code in core](https://github.com/stellar/stellar-core/blob/master/src/util/types.cpp#L184-L210)? I'm not sure if we have any in soroban, it's a few layers up from code I'm familiar with.

Reply:
I'll run a test of this to verify if we have an inconsistency in Soroban and open an issue if so. Thanks!